consensus/src/quorum_store/proof_manager.rs

```diff
@@ -0,0 +1,182 @@
+use futures_channel::mpsc::Receiver;
+use std::collections::HashSet;
+
+#[derive(Debug)]
+pub enum ProofManagerCommand {
+    LocalProof(ProofOfStore),
+    RemoteProof(ProofOfStore),
+    CommitNotification(LogicalTime, Vec<HashValue>),
+    Shutdown(tokio::sync::oneshot::Sender<()>),
+}
+
+pub struct ProofManager {
+    proofs_for_consensus: ProofQueue,
+    latest_logical_time: LogicalTime,
+    back_pressure_local_batch_limit: usize,
+    remaining_local_proof_num: usize,
+}
+
+impl ProofManager {
+    pub fn new(epoch: u64, back_pressure_local_batch_limit: usize) -> Self {
+        Self {
+            proofs_for_consensus: ProofQueue::new(),
+            latest_logical_time: LogicalTime::new(epoch, 0),
+            back_pressure_local_batch_limit,
+            remaining_local_proof_num: 0,
+        }
+    }
+
+    pub(crate) async fn handle_local_proof(
+        &mut self,
+        proof: ProofOfStore,
+        network_sender: &mut NetworkSender,
+    ) {
+        self.proofs_for_consensus.push(proof.clone(), true);
+        network_sender.broadcast_proof_of_store(proof).await;
+    }
+
+    pub(crate) fn handle_remote_proof(&mut self, proof: ProofOfStore) {
+        self.proofs_for_consensus.push(proof, false);
+    }
+
+    pub(crate) fn handle_commit_notification(
+        &mut self,
+        logical_time: LogicalTime,
+        digests: Vec<HashValue>,
+    ) {
+        debug!("QS: got clean request from execution");
+        assert_eq!(
+            self.latest_logical_time.epoch(),
+            logical_time.epoch(),
+            "Wrong epoch"
+        );
+        assert!(
+            self.latest_logical_time <= logical_time,
+            "Decreasing logical time"
+        );
+        self.latest_logical_time = logical_time;
+        self.proofs_for_consensus.mark_committed(digests);
+    }
+
+    pub(crate) fn handle_proposal_request(&mut self, msg: BlockProposalCommand) {
+        match msg {
+            // TODO: check what max_txns consensus is using
+            BlockProposalCommand::GetBlockRequest(round, max_txns, max_bytes, filter, callback) => {
+                // TODO: Pass along to batch_store
+                let excluded_proofs: HashSet<HashValue> = match filter {
+                    PayloadFilter::Empty => HashSet::new(),
+                    PayloadFilter::DirectMempool(_) => {
+                        unreachable!()
+                    },
+                    PayloadFilter::InQuorumStore(proofs) => proofs,
+                };
+
+                let proof_block = self.proofs_for_consensus.pull_proofs(
+                    &excluded_proofs,
+                    LogicalTime::new(self.latest_logical_time.epoch(), round),
+                    max_txns,
+                    max_bytes,
+                );
+                self.remaining_local_proof_num = self
+                    .proofs_for_consensus
+                    .clean_local_proofs(LogicalTime::new(self.latest_logical_time.epoch(), round));
```

Comment:
@danielxiangzl it seems redundant and premature to clean at this point. I'll remove this and rely only on the commit - let me know if it seems wrong.

Reply:
Yes it looks redundant. Basically, the commit also happens when there is a proposal.